ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: optimal token ownership must be a positive multiple of number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorLowerAndUpperBoundModulo = func(zonesCount, optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: lower and upper bounds must be congruent modulo number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorDistanceBetweenTokensNotBigEnough = func(optimalTokenOwnership, ownership int, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: distance between lower and upper bound %d is not big enough", token.prevToken, token.token, optimalTokenOwnership, ownership)
+	}
+)
+
+type SpreadMinimizingConfig struct {
+	InstanceID        string
+	Zone              string
+	TokensPerInstance int
+}
+
+type SpreadMinimizingTokenGenerator struct {
+	cfg        SpreadMinimizingConfig
+	instanceID int
+	zoneID     int
+	zones      []string
+	logger     log.Logger
+}
+
+func NewSpreadMinimizingTokenGenerator(cfg SpreadMinimizingConfig, zones []string, logger log.Logger) (*SpreadMinimizingTokenGenerator, error) {
+	sortedZones := make([]string, len(zones))
+	copy(sortedZones, zones)
+	if !slices.IsSorted(sortedZones) {
+		sort.Strings(sortedZones)
+	}
+	instanceID, err := getInstanceID(cfg.InstanceID)
+	if err != nil {
+		return nil, err
+	}
+	zoneID, err := getZoneID(cfg.Zone, sortedZones)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenerator := &SpreadMinimizingTokenGenerator{
+		cfg:        cfg,
+		instanceID: instanceID,
+		zoneID:     zoneID,
+		zones:      sortedZones,
+		logger:     logger,
+	}
+	return tokenGenerator, nil
+}
+
+func getInstanceID(instanceID string) (int, error) {
+	parts := instanceIDRegex.FindStringSubmatch(instanceID)
+	if len(parts) != 4 {
+		return 0, errorBadInstanceIDFormat(instanceID)
+	}
+	return strconv.Atoi(parts[3])
+}
+
+// getZoneID gets a zone name and a slice of sorted zones,
+// and return the index of the zone in the slice.
+func getZoneID(zone string, sortedZones []string) (int, error) {
+	if !slices.IsSorted(sortedZones) {
+		slices.Sort(sortedZones)
+	}
+	index := sort.SearchStrings(sortedZones, zone)
+	if index >= len(sortedZones) {
+		return 0, errorZoneNotValid(zone)
+	}
+	return index, nil
+}
+
+// generateFirstInstanceTokens calculates a set of tokens that should be assigned to the first instance (with id 0)
+// of the zone of the underlying instance.
+func (t *SpreadMinimizingTokenGenerator) generateFirstInstanceTokens() Tokens {
+	zonesCount := len(t.zones)
+	tokensPerInstance := t.cfg.TokensPerInstance
+	tokenDistance := (totalTokensCount / tokensPerInstance / zonesCount) * zonesCount
```

Comment:
As discussed offline, relying on `len(t.zones)` makes it impossible to change the number of zones. Can we remove this calculation and rely on `const maxZonesCount = 5` everywhere? That would allow us scaling up to 5 zones as needed, which is a reasonable tradeoff.

Reply:
Also, as an extension to this: we're doing `/ zonesCount) * zonesCount` everywhere, with zonesCount being usually 3, sometimes 5. That's an ugly math for computers and it will always lead to strange numbers.

What if we set `const maxZonesCount = 8`, so `/ zonesCount) * zonesCount` can be translated by the compiler into `>> 3) << 3`, and also the modulo checks below would become just `& 0b111`.

This will also make it easier to understand what's assigned to each zone: each zone has $2^{32-3}$ tokens space, rather than $2^{32}/5$ (which isn't even divisible so it's harder to reason about, IMO).

I've checked few of your tests and they run fine with:
```
	zones = func() (zones []string) {
		for i := 0; i < 8; i++ {
			zones = append(zones, fmt.Sprintf("zone-%s", string([]byte{'a' + byte(i)})))
		}
		return zones
	}()
```

WDYT?